### PR TITLE
Introduce new AI Mission Type: PROTECT_REGION

### PR DESCRIPTION
**Pull request by Morlic-fo**
This PR introduces a new MissionType for the AI: PROTECT_REGION.

In this mission type, the fleet will not defend its target system passively but rather try to actively engage and intercept nearby (currently: neighboring) enemy fleets if there is no immediate threat to the (primary) target system. 

This PR also assigns the PROTECT_REGION mission to idle fleets as a filler mission (targeting the closest owned system).

### CODE REVIEW

**Comment by Dilvish-fo** [comment]
Looks like a great idea, and I like very much that you've implemented the neighbor threat check to help protect against getting lured into ambushes.  I didn't go through the connections with the regular code enough to determine just when the protection mission would end; could you summarize that for me?

**Comment by Morlic-fo** [comment]
> I like very much that you've implemented the neighbor threat check to help protect against getting lured into ambushes

That isn't actually implemented yet. For now, it will still move to the target system if there is an overwhelming threat nearby (hopefully, often enough the military allocation code was correct in assuming the fleet and planetary defenses in target system are sufficient). While avoiding ambushes is definitely a good idea, that should be implemented and aimed for all mission types not this mission type in particular and so is somewhat beyond the scope of this PR.


> I didn't go through the connections with the regular code enough to determine just when the protection mission would end; could you summarize that for me?

For now, the mission will (i.e. should) end when the last order is executed - either by arriving at the target system or at a neighboring system. The idea is to let the MilitaryAI reallocate (portions of) the fleet to the same system if necessary or abandon the system if higher priority targets must be defended.

Maybe should add additional checks to verify the threat has been neutralized...

**Comment by Morlic-fo** [comment]
There is some issue related with fleet mergers which should be adressed before merging this. Working on a fix.

**Comment by Morlic-fo** [comment]
Note that this PR resolves a huge issue with doomstacks.

A common case is where empires A and B have neighboring colonies. Each of them only wants to protect its own colony so feedback will result in them stacking up defenses but noone makes a move (as no AI ever intended to attack).
With this PR, the runaway AI will make an attack move into the neighbor system, eliminating the enemy doomstack. Afterwards, it can reuse its fleet for other missions.

Take a look at this example savegame: [example_protect_region_doomstacks_resolved.zip](https://github.com/freeorion/freeorion/files/2862137/example_protect_region_doomstacks_resolved.zip)
The doomstack situation occurs in 2 locations:
1) Sakuru alpha (1071) and Hagalaz (987)
2) Yarrow beta (868) and Yarrow gamma (1001)

With the full PR, the respectively leading AI will merge its doomstack and engage the enemy.

Not ready for merge yet - AI keeps merging and splitting fleets multiple times at this point, looking for best fix.

File: default/python/AI/EnumsAI.py [hunk 1]
```diff
@@ -102,6 +102,7 @@ class MissionType(Enum):
     ORBITAL_DEFENSE = 12
     ORBITAL_INVASION = 13
     ORBITAL_OUTPOST = 14
+    PROTECT_REGION = 15
     # ORBITAL_COLONISATION = 15 Not implemented yet
```
Comment (genestack-solomatin) [inline default/python/AI/EnumsAI.py:105]: Change value for `PROTECT_REGION` or for `ORBITAL_COLONISATION`

File: default/python/AI/MilitaryAI.py [hunk 2]
```diff
@@ -921,10 +923,46 @@ def assign_military_fleets_to_systems(use_fleet_id_list=None, allocations=None,
...
+                elif get_system_jump2_threat(_sys_id):
+                    weight = 1
+                else:
+                    weight = 1 / float(state.get_distance_to_enemy_supply(_sys_id))**1.25
+                return float(weight) / (jump_distance+1)
+
+            for fid in avail_mil_fleet_ids:
+                fleet = universe.getFleet(fid)
+                FleetUtilsAI.get_fleet_system(fleet)
+                sys_distances = sorted([(system_score(fid, _sys_id), _sys_id)
```
Comment (genestack-solomatin) [inline default/python/AI/MilitaryAI.py]:
```python
systems = state.get_empire_planets_by_system().keys()
if not systems: 
    continue
sys_id = max(systems,  key=lambda x: system_score(fid, x))

**Review by genestack-solomatin — CHANGES_REQUESTED** [review]

File: default/python/AI/AIFleetMission.py [hunk 3]
```diff
@@ -139,79 +143,53 @@ def _get_fleet_order_from_target(self, mission_type, target):
...
             return
+
+        if not self.target:
+            debug("Mission has no valid target - do not merge.")
+            return
+
         universe = fo.getUniverse()
         empire_id = fo.empireID()
+
+        # only merge PROTECT_REGION if there is any threat near target
```
[the remark this answers is not in the document]
Comment (Morlic-fo) [reply default/python/AI/AIFleetMission.py:162]: If fleet A and fleet B have different mission types or different targets, then they will not be merged.

**Comment by Morlic-fo** [comment]
Overall happy with how this plays now - ready to merge from my side.

**Comment by geoffthemedio** [comment]
@Morlic-fo I'll give @Dilvish-fo or @Cjkjvfnby a few days to comment, though don't expect any given recent inactivity.

File: default/python/AI/AIFleetMission.py [hunk 4]
```diff
@@ -139,79 +135,53 @@ def _get_fleet_order_from_target(self, mission_type, target):
...
-                            need_left = 1.5 * sum(aistate.systemStatus.get(nid, {}).get('fleetThreat', 0)
-                                                  for nid in neighbors if nid != target_id)
-                            fleet_rating = CombatRatingsAI.get_fleet_rating(fid)
-                            if need_left < fleet_rating:
-                                do_merge = True
-            if do_merge:
-                FleetUtilsAI.merge_fleet_a_into_b(fid, fleet_id, need_left,
-                                                  context="Order %s of mission %s" % (context, self))
-        return
+            if fleet_mission.type != self.type or fleet_mission.target != self.target:
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py:181]: FYI: you can replace

```python
fleet_mission.type != self.type or fleet_mission.target != self.target
``` 
with
```python
(fleet_mission.type, fleet_mission.target ) != (self.type, self.target)
```

Sometimes this approach can improve readability.

File: default/python/AI/AIFleetMission.py [hunk 5]
```diff
@@ -139,79 +135,53 @@ def _get_fleet_order_from_target(self, mission_type, target):
...
-        # merging if there is no local threat (it tends to lead to fleet object churn)
-        if self.type in COMBAT_MISSION_TYPES and len(self.orders) == 1 and not system_status.get('totalThreat', 0):
+        main_fleet_system_id = main_fleet.systemID
+        if main_fleet_system_id == INVALID_ID:
+            debug("Can't merge: fleet in middle of starlane")
             return
-        destroyed_list = list(universe.destroyedObjectIDs(empire_id))
+
+        # only merge PROTECT_REGION if there is any threat near target
+        if self.type == MissionType.PROTECT_REGION:
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py:163]: Consider adding TODO:
```
Logic about cheking if we can merge certain mission types can be moved to MissionType class itself.
```


PS.  It can be easily implemented via inheritance or via mixins.

Mixin approach:

```python

class Base:
    def mergable(self, other):
        return False
    
# has a single method!
class Mergable:
    def mergable(self, other):
        return True
    
class TypaA(Base):
    pass

class TypeB(Base, Mergable):
    pass


class TypeC(Base):
    def mergable(self, other):
        ...
```

File: default/python/AI/AIFleetMission.py [hunk 6]
```diff
@@ -564,6 +590,41 @@ def issue_fleet_orders(self):
                         MilitaryAI.assign_military_fleets_to_systems(use_fleet_id_list=new_military_fleets,
                                                                      allocations=allocations)
 
+    def _portion_of_fleet_needed_here(self):
+        """Calculate the portion of the fleet needed in target system considering enemy forces."""
+        # TODO check rating against planets
+        if assertion_fails(self.type in COMBAT_MISSION_TYPES):
+            error("%s" % self)
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py]: Can be simplified to `error(self)`.

You can add mesage to check.
`assertion_fails(self.type in COMBAT_MISSION_TYPES, str(self)):`

File: default/python/AI/AIFleetMission.py [hunk 7]
```diff
@@ -564,6 +590,41 @@ def issue_fleet_orders(self):
...
+            return 0
+        if assertion_fails(self.target and self.target.id != INVALID_ID):
+            error("%s" % self)
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py]: same here

File: default/python/AI/AIFleetMission.py [hunk 8]
```diff
@@ -660,3 +725,85 @@ def __str__(self):
                                                                             (fleet and len(fleet.shipIDs)) or 0,
                                                                             CombatRatingsAI.get_fleet_rating(fleet_id),
                                                                             self.target or 'no target')
+
+    def _get_target_for_protection_mission(self):
+        """Get a target for a PROTECT_REGION mission.
+
+        1) If primary target (system target of this mission) is under attack, move to primary target.
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py:730]: what if not strong enough?

File: default/python/AI/AIFleetMission.py [hunk 9]
```diff
@@ -660,3 +725,85 @@ def __str__(self):
...
+        2) If neighbors of primary target have local fleets weaker than this fleet, may move to attack
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py]: only if neighbors are attacked?

File: default/python/AI/AIFleetMission.py [hunk 10]
```diff
@@ -660,3 +725,85 @@ def __str__(self):
...
+        3) If no neighboring fleets or strongest enemy fleet is too strong, move to primary target
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py]: `strongest enemy fleet is too strong` should we consider all fleets?

Comment (Cjkjvfnby) [reply default/python/AI/AIFleetMission.py]: Stopped here.

File: default/python/AI/CombatRatingsAI.py [hunk 11]
```diff
@@ -291,6 +291,10 @@ def get_fleet_rating_against_planets(fleet_id):
     return FleetCombatStats(fleet_id, consider_refuel=False).get_rating_vs_planets()
 
 
+def get_ship_rating(ship_id, enemy_stats=None):
```
Comment (Cjkjvfnby) [inline default/python/AI/CombatRatingsAI.py:294]: Do we need one line function for single use?

**Comment by Cjkjvfnby** [comment]
> Overall happy with how this plays now - ready to merge from my side.

I did not check how it affects game balance, just code organization.  All my comments are optional.

File: default/python/AI/AIFleetMission.py (hunk 8 above)
Comment (Morlic-fo) [reply default/python/AI/AIFleetMission.py:730]: As is the case with current mission types, it will finally be handled by movement check code at later stage where more information is available.
